File: ConnectionState.cs
namespace Quark;

public sealed class ConnectionState
{
    private readonly object _lock = new();
    private readonly Dictionary<string, (string Label, string ConsoleId)> _usbDevices = new();
    private readonly Dictionary<string, string> _networkClients = new(); 

    public void AddUsb(string deviceId, string label, string consoleId)
    {
        lock (_lock) _usbDevices[deviceId] = (label, consoleId);
    }

    public void RemoveUsb(string deviceId)
    {
        lock (_lock) _usbDevices.Remove(deviceId);
    }

    public void AddNetwork(string ip, string consoleId)
    {
        lock (_lock) _networkClients[ip] = consoleId;
    }

    public void RemoveNetwork(string ip)
    {
        lock (_lock) _networkClients.Remove(ip);
    }

    public void RenameNetwork(string oldId, string newId)
    {
        lock (_lock)
        {
            var entry = _networkClients.FirstOrDefault(kv => kv.Value == oldId);
            if (entry.Key != null)
            {
                _networkClients[entry.Key] = newId;
            }
        }
    }

    public Snapshot GetSnapshot()
    {
        lock (_lock)
            return new Snapshot(
                _usbDevices.Values.Select(v => (v.Label, v.ConsoleId)).ToList(),
                _networkClients.Select(kv => (kv.Key, kv.Value)).ToList());
    }

    public sealed record Snapshot(
        List<(string Label, string ConsoleId)> UsbDevices,
        List<(string Ip, string ConsoleId)>    NetworkClients)
    {
        public int  UsbCount   => UsbDevices.Count;
        public int  NetCount   => NetworkClients.Count;
        public bool HasUsb     => UsbCount > 0;
        public bool HasNetwork => NetCount > 0;
        public bool IsIdle     => !HasUsb && !HasNetwork;
        public int  TotalCount => UsbCount + NetCount;
        public bool IsMulti    => TotalCount > 1;

        public string PillText()
        {
            if (IsIdle) return "Listening for connection...";

            string usbPart = UsbCount == 1
                ? $"● USB:  {UsbDevices[0].ConsoleId}"
                : UsbCount > 1
                ? $"● USB ({UsbCount})"
                : "";

            string netPart = NetCount == 1
                ? $"◈ Network:  {NetworkClients[0].ConsoleId}"
                : NetCount > 1
                ? $"◈ Network ({NetCount})"
                : "";

            if (!string.IsNullOrEmpty(usbPart) && !string.IsNullOrEmpty(netPart))
                return $"{usbPart}  {netPart}";

            return string.IsNullOrEmpty(usbPart) ? netPart : usbPart;
        }

        public string PillColor()
        {
            if (IsIdle)                 return "#607D8B";
            if (HasUsb  && !HasNetwork) return "#4CAF50";
            if (!HasUsb && HasNetwork)  return "#42A5F5";
            return "#FFB74D";
        }

        public List<string> DropdownRows()
        {
            var rows = new List<string>();
            foreach (var (label, id) in UsbDevices)
                rows.Add($"● USB:  {id} - {label}");
            foreach (var (ip, id) in NetworkClients)
                rows.Add($"◈ Network:  {id} - {ip}");
            return rows;
        }
    }
}

File: android/app/src/main/java/io/github/quark/service/ConnectionState.kt
package io.github.quark.service

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData

data class UsbEntry(val key: String, val label: String, val consoleId: String)
data class NetworkEntry(val key: String, val ip: String, val consoleId: String)

object ConnectionState {

    data class Snapshot(
        val usbDevices: List<UsbEntry>,
        val networkClients: List<NetworkEntry>,
    ) {
        val usbCount get() = usbDevices.size
        val netCount get() = networkClients.size
        val isIdle get() = usbCount == 0 && netCount == 0
        val totalCount get() = usbCount + netCount
    }

    private val lock = Any()
    private val usbMap = LinkedHashMap<String, Pair<String, String>>()
    private val netMap = LinkedHashMap<String, Pair<String, String>>()

    private val _state = MutableLiveData(Snapshot(emptyList(), emptyList()))
    val state: LiveData<Snapshot> = _state

    fun addUsb(key: String, label: String, consoleId: String) {
        synchronized(lock) { usbMap[key] = Pair(label, consoleId) }
        publish()
    }

    fun removeUsb(key: String) {
        synchronized(lock) { usbMap.remove(key) }
        publish()
    }

    fun addNetwork(key: String, ip: String, consoleId: String) {
        synchronized(lock) { netMap[key] = Pair(ip, consoleId) }
        publish()
    }

    fun removeNetwork(key: String) {
        synchronized(lock) { netMap.remove(key) }
        publish()
    }

    fun renameNetwork(oldId: String, newId: String) {
        synchronized(lock) {
            val key = netMap.entries.firstOrNull { it.value.second == oldId }?.key
            if (key != null) {
                val ip = netMap[key]!!.first
                netMap[key] = Pair(ip, newId)
            }
        }
        publish()
    }

    fun renameUsb(oldId: String, newId: String) {
        synchronized(lock) {
            val key = usbMap.entries.firstOrNull { it.value.second == oldId }?.key
            if (key != null) {
                val label = usbMap[key]!!.first
                usbMap[key] = Pair(label, newId)
            }
        }
        publish()
    }

    private fun publish() {
        synchronized(lock) {
            val usbList = usbMap.entries.map { (key, value) -> UsbEntry(key, value.first, value.second) }
            val netList = netMap.entries.map { (key, value) -> NetworkEntry(key, value.first, value.second) }
            _state.postValue(Snapshot(usbList, netList))
        }
    }
}
